actors/builtin/miner/miner_actor.go

```diff
@@ -487,6 +467,186 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
+				}
+
+				// Record sectors for proof verification
+				allSectors = append(allSectors, partition.Sectors)
+				allIgnored = append(allIgnored, partition.Faults)
+				allIgnored = append(allIgnored, partition.Terminated)
+
+				// Record active sectors for marking faults.
+				active, err := partition.ActiveSectors()
+				if err != nil {
+					return err
+				}
+				err = faults.Add(partIdx, active)
+				if err != nil {
+					return err
+				}
+
+				// Record faulty power for penalties.
+				//
+				// NOTE: This also includes power that was
+				// activated at the end of the last challenge
+				// window, and power from sectors that have since
+				// expired. That means we may end up over-penalizing,
+				// but that's fine. The miner submitted a bad proof.
+				faultyPower = faultyPower.Add(partition.ActivePower())
+				return nil
+			})
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions")
+
+			allSectorsNos, err := bitfield.MultiMerge(allSectors...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge sector bitfields")
+
+			allIgnoredNos, err := bitfield.MultiMerge(allIgnored...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge fault bitfields")
+
+			// Load sectors as seen at the end of the last proving period.
+			sectorsSnapshot, err := LoadSectors(store, dl.SectorsSnapshot)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+			sectorInfos, err := sectorsSnapshot.LoadForProof(allSectorsNos, allIgnoredNos)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors for post challenge")
+
+			// Find the proving period start for the deadline in question.
+			ppStart := st.ProvingPeriodStart
+			if st.CurrentDeadline < params.Deadline {
+				ppStart -= WPoStProvingPeriod
+			}
+
+			targetDeadline := NewDeadlineInfo(ppStart, params.Deadline, currEpoch)
+
+			// Fails if validation succeeds.
+			challengeWindowedPost(rt, targetDeadline.Challenge, sectorInfos, post.Proofs)
+
+			// Ok, so, here's the tricky part. Now we need to mark these all as _bad_. However, compaction could have
+			// happened, so the sectors could have moved.
+			//
+			// We've addressed this in two ways:
+			//
+			// 1. Compaction cannot be done for Finality epochs after a challenge window.
+			// 2. Successful challenges will always penalize, even if we can't find the target sectors and mark them faulty.
+			//    You can run but you can't hide.
+
+			// Load sectors. We can't use the snapshot as a sector
+			// with an extended expiration could end up being
+			// scheduled to expire at the wrong time.
+			//
+			// Unfortunately, this means we'll need to load all
+			// sector infos twice (once for the proof, once to mark
+			// faulty). We can't use memoized power info because
+			// sectors may have been terminated.
```

Comment:
If we somehow avoid needing to load the current sector info to handle extensions, we could possibly memoize some other information in order to avoid recalculating power. E.g. we could store the terminated power since last challenge window in each deadline/partition.

Reply:
If we make the "compaction" and "challenge" windows disjoint, we can get rid of the sectors snapshot. If we get rid of the sectors snapshot, then determining the power from the sectors is effectively "free" because we already need to load them for the proof verification.